java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/TypedValueBinder.java

```diff
@@ -0,0 +1,126 @@
+
+  public TypedValueBinder(PreparedStatement preparedStatement, BufferAllocator bufferAllocator) {
+    this.parameters = VectorSchemaRoot.create(preparedStatement.getParameterSchema(), bufferAllocator);
+    this.preparedStatement = preparedStatement;
+  }
+
+  /**
+   * Bind the given Avatica values to the prepared statement.
+   * @param typedValues The parameter values.
+   */
+  public void bind(List<TypedValue> typedValues) {
+    if (preparedStatement.getParameterSchema().getFields().size() != typedValues.size()) {
+      throw new IllegalStateException(
+          String.format("Prepared statement has %s parameters, but only received %s",
+              preparedStatement.getParameterSchema().getFields().size(),
+              typedValues.size()));
+    }
+
+    for (int i = 0; i < typedValues.size(); i++) {
+      final TypedValue param = typedValues.get(i);
+      final FieldVector vector = parameters.getVector(i);
+      bindValue(param, vector);
+    }
+
+    if (!typedValues.isEmpty()) {
+      parameters.setRowCount(1);
+      preparedStatement.setParameters(parameters);
+    }
+  }
+
+  private void bindValue(TypedValue param, FieldVector vector) {
+    Object value = param.value;
+    try {
+      if (vector instanceof IntVector) {
+        ((IntVector) vector).setSafe(0, (int) value);
+      } else if (vector instanceof BigIntVector) {
+        BigIntVector longVec = (BigIntVector) vector;
+        if (value instanceof Long) {
+          longVec.setSafe(0, (long) value);
+        } else {
+          longVec.setSafe(0, (int) value);
+        }
+      } else if (vector instanceof BitVector) {
+        ((BitVector) vector).setSafe(0, (int) value);
+      } else if (vector instanceof Float4Vector) {
+        ((Float4Vector) vector).setSafe(0, (float) value);
+      } else if (vector instanceof Float8Vector) {
+        ((Float8Vector) vector).setSafe(0, (double) value);
+      } else if (vector instanceof DecimalVector) {
+        ((DecimalVector) vector).setSafe(0, (BigDecimal) value);
+      } else if (vector instanceof VarCharVector) {
+        ((VarCharVector) vector).setSafe(0, new Text((String) value));
+      } else if (vector instanceof LargeVarCharVector) {
+        ((LargeVarCharVector) vector).setSafe(0, new Text((String) value));
+      } else if (vector instanceof DateMilliVector) {
+        String pattern = "yyyy-MM-dd HH:mm:ss.SSS";
+        DateTime dateTime = DateTime.parse((String) value, DateTimeFormat.forPattern(pattern));
+
+        ((DateMilliVector) vector).setSafe(0, dateTime.getMillis());
+      } else if (vector instanceof VarBinaryVector) {
+        ((VarBinaryVector) vector).setSafe(0, (byte[]) value);
+      } else if (vector instanceof LargeVarBinaryVector) {
+        ((LargeVarBinaryVector) vector).setSafe(0, (byte[]) value);
+      } else if (vector instanceof NullVector) {
+        assert true;
+      } else {
+        throw new UnsupportedOperationException(
```

Comment:
Complex types such as Map, list, and struct aren't being covered here. We should report a different error for types that aren't yet implemented vs. types that are plain unrecognized.

Reply:
Fair point!